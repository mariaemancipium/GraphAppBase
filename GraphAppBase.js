import React, { Component } from "react";
import * as PropTypes from "prop-types";
import { shallowEqual } from "../../lib/utils";
import { DriverProvider } from "../DriverProvider";
import { DesktopIntegration } from "../DesktopIntegration";
import { connectDriver, getActiveDatabaseCredentials } from "./helpers";

export const CONNECTED = "CONNECTED";
export const DISCONNECTED = "DISCONNECTED";

export class GraphAppBase extends Component {
  constructor(props) {
    super(props);
    this.driver = {};
    this.listeners = {};
  }
  state = {
    driverCredentials: null,
    connectionState: DISCONNECTED,
    connectionDetails: null,
    initialDesktopContext: null,
    retry: 0
  };
  shouldComponentUpdate(props, state) {
    return !(
      state.connectionState === this.state.connectionState &&
      state.retry === this.state.retry &&
      shallowEqual(state.connectionDetails, this.state.connectionDetails)
    );
  }
  componentDidCatch(e) {}
  onDiMount = context => {
    this.setState({ initialDesktopContext: context });
    this.onConnectionChange(context);
  };
  onConnectionChange = context => {
    const driverCredentials = getActiveDatabaseCredentials(context);
    driverCredentials.username =
      driverCredentials.username ||
      (this.state.driverCredentials && this.state.driverCredentials.username);
    driverCredentials.password =
      driverCredentials.password ||
      (this.state.driverCredentials && this.state.driverCredentials.password);
    this.setState({ driverCredentials }, this.connectDriver);
  };
  connectDriver = () => {
    if (this.driver && this.driver.close) this.driver.close();
    const success = driver => {
      this.driver = driver;
      this.setState({
        connectionState: CONNECTED,
        connectionDetails: null
      });
    };
    connectDriver(
      this.state.driverCredentials,
      this.props.driverFactory,
      success,
      this.handleDriverError
    );
  };

  handleDriverError = e => {
    this.setState(state => ({
      connectionState: DISCONNECTED,
      connectionDetails: e,
      retry: state.retry + 1
    }));
  };
  setCredentials = (username, password) => {
    const driverCredentials = {
      ...this.state.driverCredentials,
      username,
      password
    };
    this.setState({ driverCredentials }, this.connectDriver);
  };
  on = (type, fn) => {
    if (!this.listeners[type]) this.listeners[type] = [];
    this.listeners[type].push(fn);
  };
  off = (type, fn) => {
    if (!this.listeners[type]) return;
    const index = this.listeners[type].indexOf(fn);
    if (index < 0) return;
    this.listeners[type].splice(index, 1);
  };
  handleEvents = (typeObj, newContext, oldContext) => {
    const { type } = typeObj;
    if (this.listeners[type]) {
      this.listeners[type].forEach(fn => fn(type, newContext, oldContext));
    }
  };
  render() {
    const {
      connectionState,
      connectionDetails,
      initialDesktopContext
    } = this.state;
    const { setCredentials, handleEvents, on, off } = this;
    return (
      <div>
        <DesktopIntegration
          integrationPoint={this.props.integrationPoint}
          onMount={this.onDiMount}
          onGraphActive={(_, context) => this.onConnectionChange(context)}
          on={handleEvents}
        />
        <DriverProvider driver={this.driver}>
          {this.props.render({
            connectionState,
            connectionDetails,
            setCredentials,
            on,
            off,
            initialDesktopContext
          })}
        </DriverProvider>
      </div>
    );
  }
}

GraphAppBase.propTypes = {
  driverFactory: PropTypes.object.isRequired,
  render: PropTypes.func.isRequired,
  integrationPoint: PropTypes.object
};

export default GraphAppBase;
