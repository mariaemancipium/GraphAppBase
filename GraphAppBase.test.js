import React from "react";
import TestRenderer from "react-test-renderer";
import {
  mockDriver,
  flushPromises,
  desktopApiContexts
} from "../../../config/test_helpers";
import { helpers as integrationHelpers } from "../DesktopIntegration";

import { GraphAppBase, CONNECTED, DISCONNECTED } from "./GraphAppBase";

const driverFactory = driver => ({
  auth: { basic: (u, p) => `${u}:${p}` },
  driver
});

test("GraphAppBase renders", () => {
  const out = TestRenderer.create(
    <GraphAppBase
      driverFactory={driverFactory()}
      render={() => "Hello from render prop"}
    />
  );
  expect(out.toJSON()).toMatchSnapshot();
});

test("GraphAppBase passes connection state and connection details to render prop", () => {
  // Given
  const integrationPoint = {
    getContext: () => Promise.resolve(desktopApiContexts.activeGraph)
  };
  const render = jest.fn(() => "Hello from render props (spy)");
  const driverMock = mockDriver();
  const driver = jest.fn(() => driverMock);

  // When
  const out = TestRenderer.create(
    <GraphAppBase
      driverFactory={driverFactory(driver)}
      render={render}
      integrationPoint={integrationPoint}
    />
  );

  // Then
  expect(render).toHaveBeenCalledTimes(1);
  expect(render).toHaveBeenLastCalledWith(
    expect.objectContaining({
      connectionState: DISCONNECTED,
      connectionDetails: null
    })
  );

  return flushPromises().then(() => {
    expect(out.toJSON()).toMatchSnapshot();
    expect(driver).toHaveBeenCalledTimes(1);
    expect(render).toHaveBeenCalledTimes(2);
    expect(render).toHaveBeenLastCalledWith(
      expect.objectContaining({
        connectionState: CONNECTED,
        connectionDetails: null
      })
    );

    // When
    driverMock.onError({ message: "no!" });

    // Then
    expect(render).toHaveBeenCalledTimes(3);
    expect(render).toHaveBeenLastCalledWith(
      expect.objectContaining({
        connectionState: DISCONNECTED,
        connectionDetails: { message: "no!" }
      })
    );
  });
});

test("GraphAppBase passes the context object to render prop", () => {
  // Given
  const integrationPoint = {
    getContext: () => Promise.resolve(desktopApiContexts.activeGraph)
  };
  const render = jest.fn(() => "Hello from render props (spy)");
  const driverMock = mockDriver();
  const driver = jest.fn(() => driverMock);

  // When
  TestRenderer.create(
    <GraphAppBase
      driverFactory={driverFactory(driver)}
      render={render}
      integrationPoint={integrationPoint}
    />
  );

  // Then
  expect(render).toHaveBeenCalledTimes(1);

  return flushPromises().then(() => {
    expect(render).toHaveBeenLastCalledWith(
      expect.objectContaining({
        initialDesktopContext: desktopApiContexts.activeGraph
      })
    );
  });
});

test("GraphAppBase exposes 'setCredentials' and tries to connect when called", () => {
  // Given
  const integrationPoint = {
    getContext: () => Promise.resolve(desktopApiContexts.activeGraph)
  };
  let outerSetCredentials;
  const render = jest.fn(({ setCredentials }) => {
    outerSetCredentials = setCredentials;
    return "In render";
  });
  const driverMock = mockDriver();
  const driver = jest.fn(() => driverMock);
  const factoryDriver = driverFactory(driver);
  const creds = integrationHelpers.getActiveCredentials(
    "bolt",
    desktopApiContexts.activeGraph
  );
  const host = `bolt://${creds.host}:${creds.port}`;
  const encrypted = creds.tlsLevel === "REQUIRED";
  const username = "Stella";
  const password = "password";

  // When
  TestRenderer.create(
    <GraphAppBase
      driverFactory={factoryDriver}
      render={render}
      integrationPoint={integrationPoint}
    />
  );

  // Then
  expect(render).toHaveBeenCalledTimes(1);
  expect(render).toHaveBeenLastCalledWith(
    expect.objectContaining({
      setCredentials: expect.anything()
    })
  );

  // When
  outerSetCredentials(username, password);

  // Then
  return flushPromises().then(() => {
    expect(driver).toHaveBeenCalledTimes(2);
    expect(driver).toHaveBeenLastCalledWith(
      host,
      factoryDriver.auth.basic(username, password),
      { encrypted }
    );
  });
});

test("GraphAppBase exposes 'on' so we can listen on events", () => {
  // Given
  let componentOnContextUpdate;
  const integrationPoint = {
    getContext: () => Promise.resolve(desktopApiContexts.activeGraph),
    onContextUpdate: fn => (componentOnContextUpdate = fn)
  };
  class MyApp extends React.Component {
    componentDidMount() {
      this.props.on(type, onContextUpdate); // Listen
      this.props.on(type2, onContextUpdate); // on two event types
    }
    render() {
      return "yo!";
    }
  }
  const onContextUpdate = jest.fn();
  const type = "EVENT";
  const type2 = "EVENT2";
  const oldContext = "newContext";
  const oldContext2 = "newContext2";
  const newContext = "newContext";
  const newContext2 = "newContext2";
  const render = jest.fn(({ on }) => <MyApp on={on} />);
  const driver = jest.fn(() => mockDriver());

  // When
  TestRenderer.create(
    <GraphAppBase
      driverFactory={driverFactory(driver)}
      render={render}
      integrationPoint={integrationPoint}
    />
  );

  // Then
  expect(render).toHaveBeenCalledTimes(1);
  expect(render).toHaveBeenLastCalledWith(
    expect.objectContaining({
      on: expect.anything()
    })
  );

  // When
  componentOnContextUpdate({ type }, newContext, oldContext);

  // Then
  return flushPromises().then(() => {
    expect(driver).toHaveBeenCalledTimes(1);
    expect(onContextUpdate).toHaveBeenCalledTimes(1);
    expect(onContextUpdate).toHaveBeenLastCalledWith(
      type,
      newContext,
      oldContext
    );

    // When
    componentOnContextUpdate({ type: type2 }, newContext2, oldContext2);
    expect(onContextUpdate).toHaveBeenCalledTimes(2);

    // Then
    return flushPromises().then(() => {
      expect(driver).toHaveBeenCalledTimes(1);
      expect(onContextUpdate).toHaveBeenCalledTimes(2);
      expect(onContextUpdate).toHaveBeenLastCalledWith(
        type2,
        newContext2,
        oldContext2
      );
    });
  });
});

test("GraphAppBase exposes 'off' so we can stop listening on events", () => {
  // Given
  let componentOnContextUpdate;
  const integrationPoint = {
    getContext: () => Promise.resolve(desktopApiContexts.activeGraph),
    onContextUpdate: fn => (componentOnContextUpdate = fn)
  };
  const onContextUpdate = jest.fn();
  const type = "EVENT";
  const type2 = "EVENT2";
  const oldContext = "newContext";
  const newContext = "newContext";
  const render = jest.fn(({ on, off }) => {
    on(type, onContextUpdate); // Listen on one event
    on(type2, onContextUpdate);
    off(type2, onContextUpdate); // Stop listening on second
    return "yo!";
  });
  const driver = jest.fn(() => mockDriver());

  // When
  TestRenderer.create(
    <GraphAppBase
      driverFactory={driverFactory(driver)}
      render={render}
      integrationPoint={integrationPoint}
    />
  );

  // Then
  expect(render).toHaveBeenCalledTimes(1);
  expect(render).toHaveBeenLastCalledWith(
    expect.objectContaining({
      off: expect.anything()
    })
  );

  // When
  componentOnContextUpdate({ type }, oldContext, newContext);

  // Then
  return flushPromises().then(() => {
    expect(driver).toHaveBeenCalledTimes(1);
    expect(onContextUpdate).toHaveBeenCalledTimes(1);
    expect(onContextUpdate).toHaveBeenLastCalledWith(
      type,
      newContext,
      oldContext
    );

    // When
    componentOnContextUpdate({ type: type2 }, null, null);

    // Then
    return flushPromises().then(() => {
      expect(driver).toHaveBeenCalledTimes(1);
      expect(onContextUpdate).toHaveBeenCalledTimes(1);
    });
  });
});
