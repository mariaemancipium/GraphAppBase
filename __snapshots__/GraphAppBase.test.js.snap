// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphAppBase passes connection state and connection details to render prop 1`] = `
<div>
  Hello from render props (spy)
</div>
`;

exports[`GraphAppBase renders 1`] = `
<div>
  Hello from render prop
</div>
`;
